tough/src/lib.rs

```diff
@@ -114,54 +121,55 @@ impl From<ExpirationEnforcement> for bool {
 /// ## Basic usage:
 ///
 /// ```rust
-/// # use std::fs::File;
 /// # use std::path::PathBuf;
 /// # use tough::RepositoryLoader;
 /// # use url::Url;
 /// # let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests").join("data").join("tuf-reference-impl");
 /// # let root = dir.join("metadata").join("1.root.json");
 /// # let metadata_base_url = Url::from_file_path(dir.join("metadata")).unwrap();
 /// # let targets_base_url = Url::from_file_path(dir.join("targets")).unwrap();
+/// # tokio_test::block_on(async {
 ///
 /// let repository = RepositoryLoader::new(
-///     File::open(root).unwrap(),
+///     &tokio::fs::read(root).await.unwrap(),
 ///     metadata_base_url,
 ///     targets_base_url,
 /// )
 /// .load()
+/// .await
 /// .unwrap();
 ///
+/// # });
 /// ```
 ///
 /// ## With optional settings:
 ///
 /// ```rust
-/// # use std::fs::File;
 /// # use std::path::PathBuf;
 /// # use tough::{RepositoryLoader, FilesystemTransport, ExpirationEnforcement};
 /// # use url::Url;
 /// # let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests").join("data").join("tuf-reference-impl");
 /// # let root = dir.join("metadata").join("1.root.json");
 /// # let metadata_base_url = Url::from_file_path(dir.join("metadata")).unwrap();
 /// # let targets_base_url = Url::from_file_path(dir.join("targets")).unwrap();
+/// # tokio_test::block_on(async {
 ///
 /// let repository = RepositoryLoader::new(
-///     File::open(root).unwrap(),
+///     &tokio::fs::read(root).await.unwrap(),
 ///     metadata_base_url,
 ///     targets_base_url,
 /// )
 /// .transport(FilesystemTransport)
 /// .expiration_enforcement(ExpirationEnforcement::Unsafe)
 /// .load()
+/// .await
 /// .unwrap();
 ///
+/// # });
 /// ```
 #[derive(Debug, Clone)]
-pub struct RepositoryLoader<R>
-where
-    R: Read,
-{
-    root: R,
+pub struct RepositoryLoader<'a> {
+    root: &'a [u8],
```

Comment:
Since the root is supposed to be "shipped with your software", I figured having it in a file or in memory would be the most common case. Having `Read`/`Stream`-like access to a file doesn't make much sense as there are no async interfaces for that in `serde_json`. So it has to be completely in memory anyway.

You can read a file into memory with a single call already, so that would stay the same with your approach in terms of convenience.
If users keep the root in memory already (because it's linked in the binary or because they get it by means we don't support directly), that would require an extra step now.

But maybe `from_https` is a common thing to do?
Also, your proposal adds the slight benefit of an early deserialization error.